Vignettes (#335)
insightsengineering/teal.reporter: vignettes/teal-report-class.Rmd

This vignette shows you how to use `teal_report` and `teal_card` to enhance your teal modules with reporting capabilities.

## Creating a teal_report

Creating a `teal_report` is straightforward - you can include your data and it automatically tracks the code needed to reproduce your analysis:

```{r basic-usage}
library(teal.reporter)

# Create a teal_report with data
report <- teal_report(
  iris = iris,
  mtcars = mtcars
)

# You can also start with an empty report
empty_report <- teal_report()
```

## Working with teal_card

The `teal_card` is where your report content lives. You can create cards with text, code, results, and any other content:

```{r teal-card-basics}
# Create a simple card
card <- teal_card(
  "## My Analysis",
  "This analysis looks at the iris dataset.",
  summary(iris)
)

# Add more content to an existing card
card <- c(card, list(
  "### Conclusion", 
  "The data shows interesting patterns."
))

# Add the card to a report
report <- teal_report(iris = iris)
teal_card(report) <- card

> ```

Reviewer: Maybe we should skip this and create a separate vignette for this object? Reason is that sections here are:
- <teal_report> creation
- <teal_card> management
- <teal_report> with eval_code

I think that teal_card section in between could distract from diving into details of teal_report. I would replace **Working with teal_card** by **Composing document with teal_report**. Which gives a following sections:
- Object initialization
- Composing a document
- Evaluating arbitrary code

I'd create a vignette based on `teal_report` object only. WDYT?

Author: Currently simplified the report so that `teal_card` is used at the end, only to show you can add text commentary

## Adding Computed Results with eval_code

Since `teal_report` inherits from `teal_data`, you can use `eval_code()` to add computed results that are automatically tracked for reproducibility:

```{r eval-code-example}
library(teal.code)

# Start with a teal_report
report <- teal_report(iris = iris)

# Add computed results using eval_code

> report <- eval_code(report, "
>   # Compute summary statistics
>   iris_summary <- summary(iris)
>   
>   # Create a simple model
>   model <- lm(Sepal.Length ~ Sepal.Width, data = iris)
>   model_summary <- summary(model)
> ")

Reviewer: Somebody could be mislead that `eval_code` can contain markdown syntax (comments which looks like heading in this context).
```suggestion
report <- eval_code(report, "
  iris_summary <- summary(iris)
  model <- lm(Sepal.Length ~ Sepal.Width, data = iris)
  model_summary <- summary(model)
")
```

---
title: "teal_report Class"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{teal_report Class}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
library(teal.reporter)
```

## Introduction

> The `teal_report` class in `teal.reporter` provides a powerful way to create reproducible reports within teal applications. 

Reviewer: Don't be too cocky about our own stuff. 
```suggestion
The `teal_report` class in `teal.reporter` provides a way to create reproducible reports by adding, editing and removing document elements.
```

Author: : )

The `teal_report` class is built on top of [`teal_data`](https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html), inheriting all its reproducibility and code-tracking capabilities while adding reporting-specific functionality through `teal_card`.

This vignette shows you how to use `teal_report` and `teal_card` to enhance your teal modules with reporting capabilities.

## Creating a teal_report

> Creating a `teal_report` is straightforward - you can include your data and it automatically tracks the code needed to reproduce your analysis:

Reviewer: When one includes data in the `teal_report()` constructor then a reproducibility will probably be incomplete.

Author: yeah, maybe lets rewrite to `create the object and start creating data with eval_code inside it`

Author: What about:

```r

A `teal_report` is an object where developers can add, edit and remove various content (markdown text, plots, tables), evaluate and add code chunks within its environment. It provides a framework for building reproducible reports by combining content management with automatic code tracking.

To ensure complete reproducibility, it's recommended to start with an empty `teal_report` and build up your data and analysis using `eval_code()`:

```

Reviewer: Very nice. I would just remove "various" from -> "... edit and remove ~~various~~ content...".

Author: ok, adjusted here https://github.com/insightsengineering/teal.reporter/pull/335/commits/e0d2d42b057a05e2984de7dcf64bde9439d558c2
thanks for the help

  
  # Create a simple model
  model <- lm(Sepal.Length ~ Sepal.Width, data = iris)
  model_summary <- summary(model)
")

# The code is automatically tracked
cat(get_code(report))

# Create a card with the results
results_card <- teal_card(
  "## Statistical Analysis",
  "### Summary Statistics",
  code_chunk("print(iris_summary)", echo = TRUE),
  "### Linear Model Results", 
  paste("**R-squared:**", round(report[["model_summary"]]$r.squared, 3)),
  "### Reproducible Code",
  code_chunk(get_code(report), echo = TRUE, eval = FALSE)
)

# Add to the report
teal_card(report) <- results_card
```

## Simple Regression Module Example

Here's a practical example of how to create a simple teal module that uses `teal_report` for reporting:

```{r simple-module}
library(teal)
library(teal.reporter)

tm_simple_regression <- function(label = "Simple Regression") {
  module(
    label = label,
    server = function(input, output, session, data, reporter, ...) {
      
      # Create the plot
      output$plot <- renderPlot({
        req(input$x_var, input$y_var)
        dataset <- data()[["iris"]]
        plot(dataset[[input$x_var]], dataset[[input$y_var]])
      })
      

>       report_data <- reactive({
>         obj <- data()
>         teal.reporter::teal_card(obj) <- 
>           c(teal.reporter::teal_card(obj), "# Module's computation")
>         eval_code(
>           obj, 
>           sprintf(
>             "correlation <- cor(iris[['%s']], iris[['%s']]);correlation", 
>             input$x_var, 
>             input$y_var
>           ),
>           keep_output = TRUE
>         )
>       })
>       report_data
>     },
>     
>     ui = function(id) {
>       ns <- NS(id)
>       fluidPage(
>         h3("Simple Regression Plot"),
>         fluidRow(
>           column(6,
>             selectInput(ns("x_var"), "X Variable:", 
>                        choices = names(iris)[1:4], selected = "Sepal.Length")
>           ),
>           column(6,
>             selectInput(ns("y_var"), "Y Variable:", 
>                        choices = names(iris)[1:4], selected = "Sepal.Width")
>           )
>         ),
>         plotOutput(ns("plot"))
>       )
>     }
>   )
> }
> ```
> 
> ## Using the Module in a teal App
> 
> Here's how you would use this module in a complete teal application:
> 
> ```{r complete-app}
> library(teal)
> library(teal.reporter)
> 
> app <- init(
>   data = teal_data(iris = iris),
>   modules = modules(
>     tm_simple_regression("Regression Analysis")
>   ),
>   header = "Simple Teal App with Reporting"
> )
> 
> # Run the app
> if (interactive()) {
>   shinyApp(app$ui, app$server)
> }
> ```
> 

Reviewer: Reverse dependency on `teal`. This content should be in `teal`
```suggestion
```

Author: yeah, make sense - I was a bit confused on where to put which content

This vignette shows you how to use `teal_report` to enhance your teal modules with reporting capabilities.

## Creating a teal_report

A `teal_report` is an object where developers can add, edit and remove various content (markdown text, plots, tables), evaluate and add code chunks within its environment. It provides a framework for building reproducible reports by combining content management with automatic code tracking.

To ensure complete reproducibility, it's recommended to start with an empty `teal_report` and build up your data and analysis using `eval_code()`:

```{r basic-usage}
library(teal.reporter)
report <- teal_report()
```

## Adding Computed Results with `eval_code`

Since `teal_report` inherits from `teal_data`, you can use `eval_code()` to add computed results that are automatically tracked for reproducibility:

```{r eval-code-example}
library(teal.code)

report <- eval_code(
  report, 
  "iris_head <- head(iris)",
  keep_output = TRUE
)
```

The code is automatically tracked

```{r get_code}
cat(get_code(report))
```

Also, the last output of `code` is automatically tracked, if you specify `keep_output = TRUE` in `eval_code()`.

> ## Converting to HTML

Reviewer: this section should be at the end

> Since `teal_report` inherits from `teal_data`, you can use `eval_code()` to add computed results that are automatically tracked for reproducibility:
> 
> ```{r eval-code-example}
> library(teal.code)
> 
> report <- eval_code(
>   report, 
>   "iris_head <- head(iris)",
>   keep_output = TRUE
> )
> ```
> 
> The code is automatically tracked
> 
> ```{r get_code}
> cat(get_code(report))
> ```
> 
> Also, the last output of `code` is automatically tracked, if you specify `keep_output = TRUE` in `eval_code()`.
> 
> ## Converting to HTML
> 
> The `teal_report` object can be converted to HTML using `toHTML()`.
> 
> ```{r html}
> library(tools)
> toHTML(report)
> ```
> 
> This is used in the `Reporter` class to display the report in the `Reporter previewer` module,
> but this is also helpful in displaying report object in HTML output Rmarkdown files.
> 
> ## Adding text content to the report
> 
> In most cases it is convenient to add code and the results of its evalutation to the report using `eval_code()`.
> For the scenarios where user would like to add commentary and section headers, `teal_card` can be used.
> 
> ```{r c_cards}
> teal_card(report) <- c(teal_card(report), "This analysis looks at the iris dataset.")
> ```
> 
> You can also prepend the content of the report by using `after` parameter in `append` function.
> 
> ```{r append_cards}
> teal_card(report) <- append(teal_card(report), "## My analysis", after = 0)
> ```
> 
> ```{r final_toHTML}	
> toHTML(report)
> ```
> 
> ## Key Benefits
> 
> Using `teal_report` in your modules provides several advantages:
> 
> 1. **Reproducibility**: All code is automatically captured via the underlying `teal_data` infrastructure
> 2. **Consistency**: Standardized way to create reports across modules  
> 3. **Flexibility**: Easy to add different types of content to reports
> 4. **Integration**: Works seamlessly with the teal reporter infrastructure
> 5. **Code Tracking**: Inherited `eval_code()` functionality ensures all computations are reproducible

Reviewer: This is not a precise suggestion so please don't apply this. This is just a proposition how this vignette should look like

```suggestion
## Adding content to the `teal_report`

`teal_report` object allows one to compose a reproducible document containing markdown components, code and its outputs. Use `teal_card(report)` to access and change elements of the document.
To add a new element in the `teal_card` one can use `c` method.

```{r}
teal_card(report) <- c(teal_card(report), "# Document section")
teal_card(report) <- c(teal_card(report), "Lorem ipsum dolor sit amet")
teal_card(report)
```

`teal_report` object also utilises `within` and `teal.code::eval_code()` which not only executes arbitrary code in its envrionment but also adds a code-chunk to the reproducible document

```{r}
report <- within(report, {
  a <- 1
  b <- 2
})
```

<here to explain inlude_output>

## Output teal_report

```{r}
toHTML()
```

Author: @gogonzo I think it's fine and it simplifies the length of the document for sure
